### PR TITLE
multi: Add ticket exhaustion check.

### PR DESCRIPTION
**Pull request by davecgh**
This is an alternative approach to #2090 which resolves several of the issues it had and is originally based of an idea mentioned by @matheusd in #2047.

---

This modifies the mining template creation logic to prevent creation of templates that would result in an unrecoverable chain     due to inevitable ticket exhaustion.

It is split into two individual commits as follows:

The first commit adds a new function named `checkTicketExhaustion` to `blockchain` which will return a new rule error if extending a given block with another block that contains a specified number of tickets will result in an unrecoverable chain due to inevitable ticket exhaustion along with tests to ensure proper functionality.

The approach taken is such that it should be easy to use in a future consensus change gated behind a vote if that is ultimately desired, which I personally think would be a good idea.

An exported variant is also provided that takes the hash of the block to extend for use by external callers such as the mining template code.

The second commit modifies the mining template creation logic to make use of the new ability along with returning an associated new mining rule error.

### CODE REVIEW

**Comment by davecgh** [comment]
Testing this might be a little bit challenging for the uninitiated, so here is a guide:


1. Modify the [simnet harness](https://github.com/decred/dcrd/blob/bf7072fb6583a7c44704897c9cc129346419266f/contrib/dcr_tmux_simnet_setup.sh) to disable automatic ticket purchases
```diff
diff --git a/contrib/dcr_tmux_simnet_setup.sh b/contrib/dcr_tmux_simnet_setup.sh
index bdc463a5..78fc11d8 100755
--- a/contrib/dcr_tmux_simnet_setup.sh
+++ b/contrib/dcr_tmux_simnet_setup.sh
@@ -133,7 +133,7 @@ tmux resize-pane -D 15
 tmux send-keys "cd ${NODES_ROOT}/${PRIMARY_WALLET_NAME}" C-m
 tmux send-keys "echo \"${WALLET_CREATE_CONFIG}\" | dcrwallet -C ../wallet.conf --create; tmux wait-for -S ${PRIMARY_WALLET_NAME}" C-m
 tmux wait-for ${PRIMARY_WALLET_NAME}
-tmux send-keys "dcrwallet -C ../wallet.conf --enableticketbuyer --ticketbuyer.limit=10" C-m
+tmux send-keys "dcrwallet -C ../wallet.conf" C-m
 tmux select-pane -t 1
 tmux send-keys "cd ${NODES_ROOT}/${PRIMARY_WALLET_NAME}" C-m
```

2. Start the simnet harness
3. Mine to height 126 on `dcrd1`:
```
$ ./mine 94
```

4. Notice the dcrd debug log starts complaining about that extending the block with fewer than 5 ticket purchases would result in an unrecoverable chain

5) Purchase 5 tickets and notice the block template is then generated as expected **NOTE: [1]**
```bash
$ ./directtickets 5
```


**NOTE: [1]**: Unfortunately, it's not possible to use the regular `tickets` script in the harness to recover because those are purchased via a split transaction which has to be mined first and the chain can't continue in order to mine that transaction.  As a workaround, use this `directtickets` script which is a quick and dirty script I threw together specifically to test this and purchases the tickets directly without a split transaction.

<details>
<summary>directtickets script</summary>

```bash
#!/usr/bin/env bash
NUM=1
case $1 in
  ''|*[!0-9]*) ;;
  *) NUM=$1 ;;
esac

stakediff=$(./ctl getstakedifficulty | jq '.current * 1e8 | round')
utxos=$(./ctl listunspent | jq "map(select(.spendable==true and (.amount*1e8 | round)>${stakediff}))")
numutxos=$(echo $utxos | jq 'length')
[[ ${numutxos} -ge ${NUM} ]] || { echo "not enough unspend outputs"; exit 1; }

votingaddr=$(./ctl getnewaddress)
commitaddr=$(./ctl getnewaddress)
fee=2970
for i in $(seq 1 1 ${NUM}); do
  utxo=$(echo "$utxos" | jq ".[$((i - 1))]")
  txid=$(echo $utxo | jq -r '.txid')
  vout=$(echo $utxo | jq -r '.vout')
  tree=$(echo $utxo | jq -r '.tree')
  amt=$(echo $utxo | jq -r '.amount*1e8 | round')
  changeamt=$(jq -rn "${amt}-${stakediff}-${fee}")
  ./ctl createrawsstx "[{\"txid\":\"${txid}\",\"vout\":${vout},\"tree\":${tree},\"amt\":${amt}}]" \
    "{\"${votingaddr}\": ${stakediff}}" \
    "[{\"addr\":\"${commitaddr}\",\"commitamt\":${stakediff},\"changeaddr\":\"${commitaddr}\",\"changeamt\":${changeamt}}]" | \
    ./ctl signrawtransaction - | jq -r .hex | \
    ./ctl sendrawtransaction -
done
```
</details>

**Review by dnldd — APPROVED** [review]

**Review by rstaudt2 — COMMENTED** [review]
Looks good, excellent test cases.  Tested locally as well per the instructions (thanks for adding that, definitely makes it easier to review).  Noticed a couple of small typos and other than that this looks good to go to me.

File: blockchain/validate_test.go [hunk 1]
```diff
@@ -1029,3 +1029,243 @@ func TestCheckConnectBlockTemplate(t *testing.T) {
...
+	)
+
+	// Create chain params based on regnet params with the specific values
+	// overridden.
+	params := chaincfg.RegNetParams()
+	params.CoinbaseMaturity = 16
+	params.TicketMaturity = 16
+	params.TicketsPerBlock = 5
+	params.StakeEnabledHeight = 32
+	params.StakeValidationHeight = 144
```
[the remark this answers is not in the document]
Comment (rstaudt2) [reply blockchain/validate_test.go]: Looks like you already made this update since I started reviewing, so can be resolved.

**Review by matheusd — COMMENTED** [review]
Verified the test data and did the functional tests.

Only found a small thing for clarification.

File: internal/mining/mining.go [hunk 2]
```diff
@@ -1893,6 +1893,14 @@ mempoolLoop:
 		}
 	}
 
+	// Ensure that mining the block would not cause the chain to become
+	// unrecoverable due to ticket exhaustion.
+	err = g.chain.CheckTicketExhaustion(&best.Hash, uint8(freshStake))
+	if err != nil {
+		log.Debug(err)
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/mining/mining.go:1900]: The idea here was that currently, no errors are logged in template creation, because there is nothing the user can do about it anyway, since the expectation is that the will only ever create valid templates.

However, this is a case (probably the only case) where the user can actually do something to resolve failure to create the template, so I thought it should be logged.

File: blockchain/validate.go [hunk 3]
```diff
@@ -3917,3 +3917,84 @@ func (b *BlockChain) CheckConnectBlockTemplate(block *dcrutil.Block) error {
...
+	stakeValidationHeight := b.chainParams.StakeValidationHeight
+	ticketMaturity := int64(b.chainParams.TicketMaturity)
+	if nextHeight+ticketMaturity+1 < stakeValidationHeight {
+		return nil
+	}
+
+	// Calculate the final pool size of live tickets after the ticket maturity
+	// period relative to the next block height.
+	//
+	// Note the the pool size in the block header indicates the pool size BEFORE
```
[the remark this answers is not in the document]
Comment (davecgh) [reply blockchain/validate.go:3949]: Fixed.

File: blockchain/validate_test.go [hunk 4]
```diff
@@ -1029,3 +1029,243 @@ func TestCheckConnectBlockTemplate(t *testing.T) {
 	g.NextBlock("b4ct", outs[3], ticketOuts[3], changeNonce)
 	acceptedBlockTemplate()
 }
+
+// TestCheckTicketExhaustion ensures the function which checks for inevitable
+// ticket exhaustion works as intended with a variety of scenarios including
+// various corner cases such as before, after, and straddling stake validation
+// height.
+func TestCheckTicketExhaustion(t *testing.T) {
+	// Hardecoded values expected by the tests so they remain valid if network
```
[the remark this answers is not in the document]
Comment (davecgh) [reply blockchain/validate_test.go:1038]: Updated.

File: blockchain/validate_test.go [hunk 5]
```diff
@@ -1029,3 +1029,243 @@ func TestCheckConnectBlockTemplate(t *testing.T) {
...
+
+	tests := []struct {
+		name        string       // test description
+		ticketInfo  []ticketInfo // num blocks and tickets to construct
+		newBlockTix uint8        // num tickets in new block for check call
+		err         error        // expected error
+	}{{
+		// Reach inevitable ticket exhaustion by not including any ticket
+		// purchases up to and including the final possible block prior to svh
+		// that can to prevent it.
```
[the remark this answers is not in the document]
Comment (davecgh) [reply blockchain/validate_test.go:1082]: Fixed.

File: internal/mining/mining.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (matheusd) [reply internal/mining/mining.go:1900]: Got it.

**Review by rstaudt2 — APPROVED** [review]

**Review by marcopeereboom — APPROVED** [review]
